// as-proto-gen/lib/file-context.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileContext = void 0;
const scope_context_1 = require("./scope-context");
const assert = __importStar(require("assert"));
class FileContext {
    constructor(generatorContext, fileDescriptor) {
        this.registeredImports = new Map();
        this.registeredDefinitions = new Set();
        this.generatorContext = generatorContext;
        this.fileDescriptor = fileDescriptor;
        this.moduleScopeContext = new scope_context_1.ScopeContext(this);
    }
    getGeneratorContext() {
        return this.generatorContext;
    }
    getFileDescriptor() {
        return this.fileDescriptor;
    }
    registerImport(importNamePath, importPath) {
        const [importName, ...importNamespace] = importNamePath.split(".");
        if (!importName) {
            throw new Error(`Cannot register empty import of ${importNamePath} from ${importPath}.`);
        }
        const importNames = this.registeredImports.get(importPath) || new Map();
        const safeImportName = importNames.get(importName) ||
            this.moduleScopeContext.registerName(importName);
        importNames.set(importName, safeImportName);
        this.registeredImports.set(importPath, importNames);
        return [safeImportName, ...importNamespace].join(".");
    }
    registerDefinition(definitionNamePath) {
        const [definitionName] = definitionNamePath.split(".");
        if (!this.registeredDefinitions.has(definitionName)) {
            // we assume that definitions are registered before imports
            assert.ok(!this.moduleScopeContext.hasRegisteredName(definitionName));
            this.registeredDefinitions.add(definitionName);
            // reserve this name
            this.moduleScopeContext.registerName(definitionName);
        }
        return definitionNamePath;
    }
    getImportsCode() {
        let importLines = [];
        for (const [importPath, importNames] of this.registeredImports) {
            const importFields = [];
            for (const [importName, safeImportName] of importNames) {
                const isAliased = importName !== safeImportName;
                importFields.push(isAliased ? `${importName} as ${safeImportName}` : `${importName}`);
            }
            importLines.push(`import { ${importFields.join(", ")} } from ${JSON.stringify(importPath)};`);
        }
        return importLines.join("\n");
    }
}
exports.FileContext = FileContext;
